components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,322 @@
+    //pub fn exists<T>(&self, ident: &T, target: &PackageTarget) -> Result<bool>
+    //where
+    //    T: Identifiable,
+    //{
+    //    let mut request = ListObjectsRequest::default();
+    //    let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+    //    request.bucket = self.bucket.to_owned();
+    //    match self.client.list_buckets().sync() {
+    //        Ok(bucket_list) => match bucket_list.buckets {
+    //            Some(bkts) => {
+    //                for bucket in bkts {
+    //                    if bucket.name.unwrap() == request.bucket {
+    //                        match self.client.list_objects(&request).sync() {
+    //                            Ok(object_list) => match object_list.contents {
+    //                                Some(contents) => {
+    //                                    for object in contents {
+    //                                        if object.key.unwrap() == key {
+    //                                            return Ok(true);
+    //                                        }
+    //                                    }
+    //                                    return Ok(false);
+    //                                }
+    //                                None => return Ok(false),
+    //                            },
+    //                            Err(e) => {
+    //                                debug!("{:?}", e);
+    //                                return Err(Error::ObjectError(e));
+    //                            }
+    //                        }
+    //                    }
+    //                }
+    //                return Ok(false);
+    //            }
+    //            None => Ok(false),
+    //        },
+    //        Err(e) => {
+    //            debug!("{:?}", e);
+    //            Err(Error::BadBucket(e))
+    //        }
+    //    }
+    //}
+
+    // This function checks whether or not the
+    // configured bucket exists in the configured
+    // backend.
+    fn exists(&self) -> Result<bool> {
+        let artifactbucket = self.bucket.to_owned();
+        match self.client.list_buckets().sync() {
+            Ok(bucket_list) => match bucket_list.buckets {
+                Some(buckets) => {
+                    for bucket in buckets {
+                        if bucket.name.unwrap() == artifactbucket {
+                            return Ok(true);
+                        } else {
+                            return Ok(false);
+                        }
+                    }
+                    return Ok(false);
+                }
+                None => Ok(false),
+            },
+            Err(e) => {
+                debug!("{:?}", e);
+                return Err(Error::ListBuckets(e));
+            }
+        }
+    }
+
+    pub fn create_bucket(&self) -> Result<()> {
+        let mut request = CreateBucketRequest::default();
+        request.bucket = self.bucket.clone();
+
+        if self.exists().unwrap() {
```

Comment:
Might be nicer to have the exists check outside of create bucket.. if !bucket_exists() { create_bucket() }.. :)

Reply:
Perhaps. For now this code path can't actually get executed. I left it in for once we decide whether or not we want on-prem users (or ourselves for that matter) to have to use an S3 account with elevated privileges. Using this functionality would require whatever service account to have both PUT, GET, and CREATE permissions in s3. Significantly smaller issue for Minio but still worth a discussion before any further work is performed here.